### ISSUE TITLE
DeepExplainer maxpooling grad error

### ISSUE THREAD

**Issue by xuesongwang** [shap/shap#2174]
Hi, 
  
I was trying to explain a convolutional network (network structure shown later) but could  not get the gradient using e.shap_values(testset). I turned on the  "with autograd.detect_anomaly():" and the error seemed to be on the maxpooling layer:

"Using a non-full backward hook when the forward contains multiple autograd Nodes is deprecated and will be removed in future versions. This hook will be missing some grad_input. Please use register_full_backward_hook to get the documented behavior.

[W python_anomaly_mode.cpp:104] Warning: Error detected in MaxPool2DWithIndicesBackward. Traceback of forward call that caused the error:
     ....
    return torch.max_pool2d(input, kernel_size, stride, padding, dilation, ceil_mode) "

![image](https://user-images.githubusercontent.com/19380806/133032757-e6b20d24-7354-4695-8649-4c00d7961658.png)

The convolutional model is defined here  (nothing really special):

    class SpatialTemporalCNN(nn.Module):

    """
    def __init__(self, input_dim=1, output_dim=2, use_aux = False):
        super(SpatialTemporalCNN, self).__init__()
        # self.l1 = nn.MaxPool2d((2, 1),1) 
        self.l1 = FactorizedConvolution(input_dim, channel_dim=16, bottle_neck=False)
        self.l2 = FactorizedConvolution(16, 32, 8)
        self.l3 = FactorizedConvolution(32,32,8)
        self.l4 = FactorizedConvolution(32,64,16)
        self.l5 = FactorizedConvolution(64,64,16)
        self.maxpool = nn.MaxPool2d(2, 2)
        self.l6 = FactorizedConvolution(64,128,32)
        self.l7 = FactorizedConvolution(128,128,32)
        self.use_aux = use_aux
        hidden_dim = self.l7.output_dim
        if use_aux == False:
            self.final_layer = nn.Linear(hidden_dim, output_dim)
        else:
            aux_dim = 10
            layers = [nn.Linear(hidden_dim+aux_dim, hidden_dim),
                      nn.ReLU(),
                      nn.Linear(hidden_dim, output_dim)]
            self.final_layer = nn.Sequential(*layers)

    def forward(self, x, return_avg = False, x_aux=None):
        """
        input x: shape (bs, 90, 16)  (batch_size, channel, num_ROI, time_step)
        """
        x = x.unsqueeze(1)  #(bs, 1, num_ROI, time_step)
        h1 = self.l1(x)
        pool1 = self.maxpool(h1)
        h2 = self.l2(pool1)
        h3 = self.l3(h2)
        pool2 = self.maxpool(h3)
        h4 = self.l4(pool2)
        h5 = self.l5(h4)
        pool3 = self.maxpool(h5)
        # avg_pool = torch.mean(h5,[2,3])  # (bs, 128)
        h6 = self.l6(pool3)
        h7 = self.l7(h6) # (bs,128, 11,2)
        # print("h7 shape", h7.shape)
        avg_pool = torch.mean(h7, [2, 3])  # (bs, 128)
        # print("avg shape", avg_pool.shape)
        if self.use_aux:
            # when in testing mode, the shape of avg_pool is different from x_aux, need to augment x_aux
            if avg_pool.size(0) != x_aux.size(0):
                aug_rate = avg_pool.size(0) // x_aux.size(0)
                x_aux = x_aux.unsqueeze(1).repeat(1, aug_rate, 1)
                x_aux = x_aux.view(avg_pool.size(0), x_aux.size(-1))
            if len(avg_pool.shape) != len(x_aux.shape):
                x_aux = x_aux[:, 0, :]
            avg_pool_cat = torch.cat([avg_pool, x_aux], dim=-1)
        output = self.final_layer(avg_pool_cat)if self.use_aux else self.final_layer(avg_pool)
        return (output, avg_pool) if (return_avg==True) else output
    
     class FactorizedConvolution(nn.Module):
     def __init__(self, input_dim = 1, output_dim =1, channel_dim = 32, bottle_neck=True):
        super(FactorizedConvolution, self).__init__()
        layers = [nn.Conv2d(input_dim,channel_dim,1,1),
                  nn.ReLU(),
                  nn.Conv2d(channel_dim, channel_dim,kernel_size=(1, 3),stride=1,padding=(0, 1)),# padding = (0, 1) to remain original shape
                  nn.ReLU(),
                  nn.Conv2d(channel_dim, channel_dim,kernel_size=(3, 1),stride=1,padding=(1, 0)),
                  nn.ReLU(),
                  nn.Conv2d(channel_dim, output_dim,1,1)
                  ]
        if bottle_neck == False:
            layers = [nn.Conv2d(input_dim, channel_dim,kernel_size=(1, 3),stride=1,padding=(0, 1)),
                      nn.ReLU(),
                      nn.Conv2d(channel_dim, channel_dim,kernel_size=(3, 1),stride=1,padding=(1, 0))
                      ]
        self.identity_match_layer = nn.Conv2d(input_dim, output_dim, 1, 1)
        self.layers = init_sequential_weights(nn.Sequential(*layers))
        self.relu = nn.ReLU()
        self.input_dim = input_dim
        self.output_dim = output_dim

    def forward(self, x):
        hidden = self.layers(x)
        if self.input_dim != self.output_dim:
            x = self.identity_match_layer(x)
        x = self.relu(x + hidden) # residual
        return x

   `

Can somebody please give any hint?  Thanks :)

**Comment by saifullah3396** [comment]
A little too late but a quick fix for this is to not reuse maxpool/relu layers. Create a separate layer for each activation. Shap has a bug that does not take reused layers into account while calculating gradients.
